lib/driver.js

```diff
@@ -442,6 +453,85 @@ class XCUITestDriver extends BaseDriver {
     // finally, return true since the superclass check passed, as did this
     return true;
   }
+  
+  async installApp () {
+    if (this.isSafari()){
+      return;
+    }
+    // if user has passed in desiredCaps.autoLaunch = false
+    // meaning they will manage app install / launching
+    if (this.opts.autoLaunch === false) {
+      return;
+    }
+
+    if (this.isRealDevice()) {
+      await this.installToRealDevice();
+    } else {
+      await this.installToSimulator();
+    }
+  }
+
+  async installToSimulator () {
+    if (this.opts.fullReset && this.opts.bundleId) {
+      await removeApp(this.opts.device.udid, this.opts.bundleId);
+    }  
+    await installApp(this.opts.device.udid, this.opts.app);
+  }
+
+  async installToRealDevice () {
+    // if we have an ipa file, set it in opts
+    if (this.opts.app) {
+      let ext = this.opts.app.substring(this.opts.app.length - 3).toLowerCase();
+      if (ext === 'ipa') {
+        this.opts.ipa = this.opts.app;
+      }
+    }
+
+    if (this.opts.udid) {
+      if (await this.opts.device.isInstalled(this.opts.bundleId)) {
+        log.debug("App is installed.");
+        if (this.opts.fullReset) {
+          log.debug("fullReset requested. Forcing app install after uninstall app");
+          await this.opts.device.remove(this.opts.bundleId);
+        } else {
+          log.debug("fullReset not requested. No need to install.");
+          return;
+        }
+      } else {
+        log.debug("App is not installed. Will try to install.");
+      }
+
+      if (this.opts.ipa && this.opts.bundleId) {
+        await this.opts.device.installAndWait(this.opts.ipa, this.opts.bundleId);
+        log.debug('App installed.');
+      } else if (this.opts.ipa) {
+        let msg = "You specified a UDID and ipa but did not include the bundle id";
+        log.warn(msg);
+        throw new errors.UnknownError(msg);
+      } else if (this.opts.app) {
+        await this.opts.device.install(this.opts.app);
+        log.debug('App installed.');
+      } else {
+        log.debug("Real device specified but no ipa or app path, assuming bundle ID is " +
+                     "on device");
+      }
+    } else {
+      log.debug("No device id or app, not installing to real device.");
+    }
+  }
+
+  getIDeviceObj () {
+    log.debug(`Creating iDevice object with udid ${this.opts.udid}`);
+    try {
+      iDevice(this.opts.udid);
+      return new Fruitsrap(this.opts.udid, {path : this.opts.fruitstrap});
+    } catch (e) {
+      let msg = "Could not initialize ideviceinstaller; make sure it is " +
```

Comment:
This message no longer applies. The error handling should probably be split so we can distinguish between an `node-idevice` error and one from (for now) Fruitstrap.